### PR TITLE
[WIP] Backup & Recovery Doc Guide

### PR DESCRIPTION
**Pull request by rhtyd**
FS: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Backup+and+Recovery+Framework
PR: https://github.com/apache/cloudstack/pull/3553

### CODE REVIEW

**Comment by rhtyd** [comment]
requesting docbuild

**Comment by rhtyd** [comment]
requesting docbuild

**Comment by rhtyd** [comment]
requesting docbuild

**Comment by rhtyd** [comment]
@PaulAngus @borisstoyanov I've submitted the feature write-up, please review/extend, edit/advise changes. Thanks.

**Comment by PaulAngus** [comment]
requesting docbuild

**Comment by PaulAngus** [comment]
requesting docbuild

**Comment by rhtyd** [comment]
Thanks @PaulAngus for the screenshots and writeup.

**Comment by DaanHoogland** [comment]
useful for learning testing. I would like to have a deployment diagram added to visualise which machines are involved (MS, VSphere, Veeam, Storage), I'll do a full review once i tested the feature.

**Comment by DaanHoogland** [comment]
tested extensively with advanced zone in isolated networks, basic feature has been tested in a shared network. Only smoke tests have been done in basic zone.

File: source/adminguide/templates.rst [hunk 1]
```diff
@@ -383,6 +385,156 @@ When templates are deleted, the VMs instantiated from them will continue
 to run. However, new VMs cannot be created based on the deleted
 template.
 
+Working with ISOs
+===================
+
```
Comment (DaanHoogland) [inline source/adminguide/templates.rst]: is this addition relevant to this pr or did it creep in?

File: source/adminguide/veeam_plugin.rst [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+(all settings can be global or per-zone)
+
+.. cssclass:: table-striped table-bordered table-hover
+
+==================================== ========================
+Configuration                         Description
+==================================== ========================
+backup.plugin.veeam.url              Veeam B&R server URL. Default: http://<VEEAM_SERVER_IP>:9399/api/
```
Comment (DaanHoogland) [inline source/adminguide/veeam_plugin.rst]: i think the default port has changed with the latest veeam: http://<VEEAM_SERVER_IP>:9398/api/

File: source/adminguide/virtual_machines.rst [hunk 3]
```diff
@@ -268,296 +247,53 @@ Once a VM instance is created, you can stop, restart, or delete it as
...
+Once a virtual machine is **expunged**, it cannot be recovered. All the
+resources used by the virtual machine will be reclaimed by the system,
+This includes the virtual machine’s IP address.  
 
+Managing Virtual Machines
+=========================
 
 Changing the VM Name, OS, or Group
-----------------------------------
+-------------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 4]
```diff
@@ -587,7 +323,7 @@ To access a VM through the CloudStack UI:
 
 
 Appending a Display Name to the Guest VM’s Internal Name
---------------------------------------------------------
+----------------------------------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 5]
```diff
@@ -622,7 +358,7 @@ No                            False                   UUID                 i-<us
 
 
 Changing the Service Offering for a VM
---------------------------------------
+----------------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 6]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
+(KVM disk controllers)
 
-   -  **OS Type**: This helps CloudStack and the hypervisor perform
-      certain operations and make assumptions that improve the
-      performance of the guest. Select one of the following.
 
-      -  If the operating system of your desired ISO image is listed,
-         choose it.
+Virtual Machine Snapshots
+==========================
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: small over-underlining

File: source/adminguide/virtual_machines.rst [hunk 7]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
+   VM snapshots are deleted automatically when a VM is destroyed. You don't 
+   have to manually delete the snapshots in this case.
 
-You can also omit the template ID parameter from the
-restoreVirtualMachine call. In this case, the VM's root disk is
-destroyed and recreated, but from the same template or ISO that was
-already in use by the VM.
 
+Virtual Machine Backups (Backup and Recovery Feature)
+======================================================
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 8]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
+
+.. include:: backup_and_recovery.rst
 
 Using SSH Keys for Authentication
----------------------------------
+===================================
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 9]
```diff
@@ -1043,7 +933,7 @@ files. Using a single SSH key pair, you can manage multiple instances.
 
 
 Creating an Instance Template that Supports SSH Keys
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+--------------------------------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 10]
```diff
@@ -1081,7 +971,7 @@ Create an instance template that supports SSH Keys.
 
 
 Creating the SSH Keypair
-~~~~~~~~~~~~~~~~~~~~~~~~
+--------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 11]
```diff
@@ -1140,7 +1030,7 @@ keypair called "keypair-doc" for the admin account in the root domain:
 
 
 Creating an Instance
-~~~~~~~~~~~~~~~~~~~~
+----------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 12]
```diff
@@ -1164,7 +1054,7 @@ environment.
 
 
 Logging In Using the SSH Keypair
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+---------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 13]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
-Adding an ISO
-~~~~~~~~~~~~~
 
-To make additional operating system or other software available for use
-with guest VMs, you can add an ISO. The ISO is typically thought of as
-an operating system image, but you can also add ISOs for other types of
-software, such as desktop applications that you want to be installed as
-part of a template.
+Advanced VM Instance Settings
+-------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: --

File: source/adminguide/virtual_machines.rst [hunk 14]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
+If you need more information about VM snapshots on VMware, check out the
+VMware documentation and the VMware Knowledge Base, especially
+`Understanding virtual machine snapshots 
+<http://kb.vmware.com/selfservice/microsites/search.do?cmd=displayKC&externalId=1015180>`_.
 
-         -  Fedora 9
 
-         -  Other PV
+Limitations on VM Snapshots
+----------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 15]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
-         -  CentOS 5.5
+-  You should use only CloudStack to create VM snapshots on hosts
+   managed by CloudStack. Any snapshots that you make directly on the
+   hypervisor will not be tracked in CloudStack.
 
-         -  Red Hat Enterprise Linux 5.3
 
-         -  Red Hat Enterprise Linux 5.4
+Configuring VM Snapshots
+--------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 16]
```diff
@@ -801,236 +537,390 @@ following conditions:
...
+vmsnapshot.create.wait Number of seconds to wait for a snapshot job to succeed before declaring failure and issuing an error.
+====================== ========================
 
-   -  **Extractable**: Choose Yes if the ISO should be available for
-      extraction.
 
-   -  **Public**: Choose Yes if this ISO should be available to other
-      users.
+Using VM Snapshots
+--------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 17]
```diff
@@ -1180,7 +1070,7 @@ The -i parameter tells the ssh client to use a ssh key found at
 
 
 Resetting SSH Keys
-~~~~~~~~~~~~~~~~~~
+--------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 18]
```diff
@@ -1192,7 +1082,7 @@ call resetSSHKeyForVirtualMachine.
 
 
 Assigning GPU/vGPU to Guest VMs
--------------------------------
+=================================
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 19]
```diff
@@ -1246,7 +1136,7 @@ CloudStack provides you with the following capabilities:
   in case of GRID cards, and capacity of the cards.
 
 Prerequisites and System Requirements
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+------------------------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 20]
```diff
@@ -1295,7 +1185,7 @@ Before continuing with configuration, consider the following:
 - Notification thresholds for GPU resource is not supported.
 
 Supported GPU Devices
-~~~~~~~~~~~~~~~~~~~~~
+------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/virtual_machines.rst [hunk 21]
```diff
@@ -1315,7 +1205,7 @@ vGPU        - GRID K100
 =========== ========================
 
 GPU/vGPU Assignment Workflow
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+-----------------------------
```
Comment (DaanHoogland) [inline source/adminguide/virtual_machines.rst]: over-underlining

File: source/adminguide/backup_and_recovery.rst [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+About Backup And Recovery
+--------------------------
+
+CloudStack version 4.14 introduces a new Backup and Recovery (B&R) framework that
+provides CloudStack with users the ability to back up their guest VMs for recovery
+purposes via 3rd party backup solutions.  The framework abstracts the API commands
+required for common backup and recovery
+operations, from the vendor specific commmands needed to perform those actions and provides
+a plugin model to enable any solution which provides backup and recovery 'like'
+features to be integrated.
```
[the remark this answers is not in the document]
Comment (PaulAngus) [reply source/adminguide/backup_and_recovery.rst:25]: 'like' because one could replicate the backup and recovery features, using a storage vendors' backend snapshot and restore capabilities,

File: source/adminguide/backup_and_recovery.rst [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+
+- VMware with Veeam Backup and Recovery
+
+See the Veeam Backup and Recovery plugin documentation for plugin specific information.
+:ref:`Veeam Backup and Recovery Plugin`
+
+Backup and Recovery Concepts
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Backup and recovery has been designed to support two modes:
```
[the remark this answers is not in the document]
Comment (PaulAngus) [reply source/adminguide/backup_and_recovery.rst:37]: The two are
1. SLA (user has no direct control)
2. Ad-hoc/user scheduled (user has direct control of backup timing)

File: source/adminguide/templates.rst (hunk 1 above)
Comment (PaulAngus) [reply source/adminguide/templates.rst:390]: i don't remember, i suspect that i found it in 'the wrong place' when structuring the additional documentation.

File: source/adminguide/veeam_plugin.rst (hunk 2 above)
Comment (PaulAngus) [reply source/adminguide/veeam_plugin.rst:132]: updated.

File: source/adminguide/virtual_machines.rst (hunk 21 above)
Comment (PaulAngus) [reply source/adminguide/virtual_machines.rst:1208]: fixed

File: source/adminguide/virtual_machines.rst (hunk 19 above)
Comment (PaulAngus) [reply source/adminguide/virtual_machines.rst:1139]: fixed

**Comment by PaulAngus** [comment]
request docbuild

**Comment by rhtyd** [comment]
@PaulAngus @andrijapanicsb should we merge this?

**Comment by andrijapanicsb** [comment]
as long as it's LGTM by Paul or Daan - I guess yes - we need it for 4.14.

**Comment by PaulAngus** [comment]
I've updated based on Daan's review so AFAIC it's good to go.

File: source/adminguide/backup_and_recovery.rst (hunk 23 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply source/adminguide/backup_and_recovery.rst:37]: ok, i can live with this. 'ad-hoc' and 'cloudstack/user scheduled' are not separate modes.
